File: MockApp/Document/DocumentViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Windows.Controls;
using SafetyProgram.Base.Interfaces;

namespace MockApp.Document
{
    public sealed class DocumentViewModel : IDocumentViewModel
    {
        public DocumentViewModel(ContextMenu contextMenu, 
            ObservableCollection<Control> content, 
            IFormat format)
        {
            if (contextMenu == null ||
                content == null ||
                format == null)
            {
                throw new ArgumentNullException();
            }
            else
            {
                this.contextMenu = contextMenu;
            }
        }

        private readonly ContextMenu contextMenu;
        public ContextMenu ContextMenu
        {
            get { return contextMenu; }
        }

        private readonly ObservableCollection<Control> content;
        public ObservableCollection<Control> Content
        {
            get { return content; }
        }

        private readonly IFormat format;
        public IFormat Format
        {
            get { return format; }
        }
    }
}

File: SafetyProgram.UI.ViewModels/DocumentViewModel.fs
namespace SafetyProgram.UI.ViewModels

open SafetyProgram.UI.Models
open SafetyProgram.Core.Models
open System.Windows.Controls
open ReactiveUI
open System

type DocumentViewModel(contentGenerator : GuiDocumentObject -> Control * IViewModel, model : GuiDocument) = 

    let propertyChanged = new Event<_,_>()

    let docObjectPairs = model.Content.CreateDerivedCollection<_,_>(Func<_,_> contentGenerator)
    let docObjectViews = docObjectPairs.CreateDerivedCollection<_,_>(Func<_,_> fst)
    
    interface IDocumentViewModel with
        member this.Width
            with get () = model.Format.Width / 0.1m<m>

        member this.Height
            with get () = model.Format.Height / 0.1m<m>

        member this.DocumentObjects
            with get () = docObjectViews

        [<CLIEvent>]
        member this.PropertyChanged = propertyChanged.Publish

        member this.Dispose() = 

            // Dispose all child viewmodels.
            docObjectPairs
            |> Seq.iter (fun (_, vm) -> vm.Dispose())

            // Dispose reactive collections linked into guimodel collection.
            docObjectViews.Dispose()
            docObjectPairs.Dispose()

    // IMPLICIT IMPL
    member this.Width
        with get () = model.Format.Width / 0.1m<m>

    member this.Height
        with get () = model.Format.Height / 0.1m<m>

    member this.DocumentObjects
        with get () = docObjectViews

    [<CLIEvent>]
    member this.PropertyChanged = propertyChanged.Publish

    member this.Dispose() = 

        // Dispose all child viewmodels.
        docObjectPairs
        |> Seq.iter (fun (_, vm) -> vm.Dispose())

        // Dispose reactive collections linked into guimodel collection.
        docObjectViews.Dispose()
        docObjectPairs.Dispose()